File: regions/system/library/dart/codegen/test/codegen_test.dart
import 'package:k1s0_codegen/codegen.dart';
import 'package:test/test.dart';

void main() {
  group('ScaffoldConfig', () {
    // 有効な設定で validate が例外をスローしないことを確認する。
    test('有効な設定でバリデーションが成功すること', () {
      const config = ScaffoldConfig(
        name: 'my-server',
        tier: Tier.system,
        apiStyle: ApiStyle.rest,
        database: DatabaseType.none,
        description: 'テスト用サーバー',
      );
      expect(() => config.validate(), returnsNormally);
    });

    // 空の名前で ConfigError がスローされることを確認する。
    test('空の名前でバリデーションが失敗すること', () {
      const config = ScaffoldConfig(
        name: '',
        tier: Tier.system,
        apiStyle: ApiStyle.rest,
        database: DatabaseType.none,
        description: 'テスト',
      );
      expect(() => config.validate(), throwsA(isA<ConfigError>()));
    });

    // 不正なケバブケースで ConfigError がスローされることを確認する。
    test('大文字を含む名前でバリデーションが失敗すること', () {
      const config = ScaffoldConfig(
        name: 'MyServer',
        tier: Tier.system,
        apiStyle: ApiStyle.rest,
        database: DatabaseType.none,
        description: 'テスト',
      );
      expect(() => config.validate(), throwsA(isA<ConfigError>()));
    });

    // 先頭ハイフンで ConfigError がスローされることを確認する。
    test('先頭にハイフンがある名前でバリデーションが失敗すること', () {
      const config = ScaffoldConfig(
        name: '-server',
        tier: Tier.system,
        apiStyle: ApiStyle.rest,
        database: DatabaseType.none,
        description: 'テスト',
      );
      expect(() => config.validate(), throwsA(isA<ConfigError>()));
    });

    // 連続ハイフンで ConfigError がスローされることを確認する。
    test('連続ハイフンがある名前でバリデーションが失敗すること', () {
      const config = ScaffoldConfig(
        name: 'my--server',
        tier: Tier.system,
        apiStyle: ApiStyle.rest,
        database: DatabaseType.none,
        description: 'テスト',
      );
      expect(() => config.validate(), throwsA(isA<ConfigError>()));
    });

    // hasGrpc / hasRest / hasDatabase の便利ゲッターが正しく動作することを確認する。
    test('便利ゲッターが正しい値を返すこと', () {
      const config = ScaffoldConfig(
        name: 'my-server',
        tier: Tier.system,
        apiStyle: ApiStyle.both,
        database: DatabaseType.postgres,
        description: 'テスト',
      );
      expect(config.hasGrpc(), isTrue);
      expect(config.hasRest(), isTrue);
      expect(config.hasDatabase(), isTrue);
    });

    // REST のみの場合に hasGrpc が false であることを確認する。
    test('RESTのみの場合にhasGrpcがfalseを返すこと', () {
      const config = ScaffoldConfig(
        name: 'my-server',
        tier: Tier.system,
        apiStyle: ApiStyle.rest,
        database: DatabaseType.none,
        description: 'テスト',
      );
      expect(config.hasGrpc(), isFalse);
      expect(config.hasRest(), isTrue);
      expect(config.hasDatabase(), isFalse);
    });
  });

  group('Tier', () {
    // displayName が正しい値を返すことを確認する。
    test('displayNameが正しい文字列を返すこと', () {
      expect(Tier.system.displayName, 'system');
      expect(Tier.business.displayName, 'business');
      expect(Tier.service.displayName, 'service');
    });
  });

  group('ApiStyle', () {
    // hasGrpc / hasRest が各値で正しく動作することを確認する。
    test('hasGrpcとhasRestが正しい値を返すこと', () {
      expect(ApiStyle.rest.hasGrpc, isFalse);
      expect(ApiStyle.rest.hasRest, isTrue);
      expect(ApiStyle.grpc.hasGrpc, isTrue);
      expect(ApiStyle.grpc.hasRest, isFalse);
      expect(ApiStyle.both.hasGrpc, isTrue);
      expect(ApiStyle.both.hasRest, isTrue);
    });
  });

  group('DatabaseType', () {
    // hasDatabase が各値で正しく動作することを確認する。
    test('hasDatabaseが正しい値を返すこと', () {
      expect(DatabaseType.postgres.hasDatabase, isTrue);
      expect(DatabaseType.none.hasDatabase, isFalse);
    });
  });

  group('Naming', () {
    // toSnakeCase がケバブケースをスネークケースに変換することを確認する。
    test('toSnakeCaseがケバブケースをスネークケースに変換すること', () {
      expect(toSnakeCase('auth-server'), 'auth_server');
      expect(toSnakeCase('my-cool-service'), 'my_cool_service');
      expect(toSnakeCase('single'), 'single');
    });

    // toPascalCase がケバブケースをパスカルケースに変換することを確認する。
    test('toPascalCaseがケバブケースをパスカルケースに変換すること', () {
      expect(toPascalCase('auth-server'), 'AuthServer');
      expect(toPascalCase('my-cool-service'), 'MyCoolService');
      expect(toPascalCase('single'), 'Single');
    });

    // toKebabCase がアンダースコア区切りをケバブケースに変換することを確認する。
    test('toKebabCaseがアンダースコア区切りをケバブケースに変換すること', () {
      expect(toKebabCase('auth_server'), 'auth-server');
      expect(toKebabCase('my_cool_service'), 'my-cool-service');
      expect(toKebabCase('single'), 'single');
    });

    // toCamelCase がケバブケースをキャメルケースに変換することを確認する。
    test('toCamelCaseがケバブケースをキャメルケースに変換すること', () {
      expect(toCamelCase('auth-server'), 'authServer');
      expect(toCamelCase('my-cool-service'), 'myCoolService');
      expect(toCamelCase('single'), 'single');
    });
  });

  group('GenerateResult', () {
    // created と skipped のリストが正しく保持されることを確認する。
    test('生成結果が正しく作成されること', () {
      const result = GenerateResult(
        created: ['file1.dart', 'file2.dart'],
        skipped: ['file3.dart'],
      );
      expect(result.created, hasLength(2));
      expect(result.skipped, hasLength(1));
      expect(result.createdCount, 2);
      expect(result.skippedCount, 1);
    });

    // 空の結果が正しく作成されることを確認する。
    test('空の生成結果が正しく作成されること', () {
      const result = GenerateResult(created: [], skipped: []);
      expect(result.createdCount, 0);
      expect(result.skippedCount, 0);
    });
  });

  group('CodegenError', () {
    // CodegenError のサブクラスが正しいコードを持つことを確認する。
    test('エラーサブクラスが正しいコードを持つこと', () {
      const configErr = ConfigError('bad config');
      expect(configErr.code, 'CONFIG_ERROR');
      expect(configErr.message, 'bad config');

      const templateErr = TemplateError('bad template');
      expect(templateErr.code, 'TEMPLATE_ERROR');

      const ioErr = IoError('file not found');
      expect(ioErr.code, 'IO_ERROR');
    });

    // CodegenError が Exception を実装していることを確認する。
    test('CodegenErrorがExceptionを実装していること', () {
      const err = CodegenError('test error');
      expect(err, isA<Exception>());
    });
  });
}

File: regions/system/library/typescript/codegen/__tests__/codegen.test.ts
import { describe, it, expect } from 'vitest';
import {
  validateConfig,
  hasGrpc,
  hasRest,
  hasDatabase,
  toSnakeCase,
  toPascalCase,
  toKebabCase,
  toCamelCase,
  CodegenError,
  type ScaffoldConfig,
  type GenerateResult,
  type ValidationResult,
} from '../src/index.js';

describe('validateConfig', () => {
  // 有効な設定でバリデーションが成功することを確認する
  it('有効な設定で valid: true を返す', () => {
    const config: ScaffoldConfig = {
      name: 'auth',
      tier: 'system',
      apiStyle: 'rest',
      database: 'postgres',
      description: '認証サーバー',
      generateClient: true,
    };
    const result = validateConfig(config);
    expect(result.valid).toBe(true);
    expect(result.errors).toHaveLength(0);
  });

  // 名前が空の場合にエラーを返すことを確認する
  it('名前が空の場合にエラーを返す', () => {
    const config: ScaffoldConfig = {
      name: '',
      tier: 'system',
      apiStyle: 'rest',
      database: 'none',
      description: 'テスト',
      generateClient: false,
    };
    const result = validateConfig(config);
    expect(result.valid).toBe(false);
    expect(result.errors).toContain('name is required');
  });

  // 名前が不正な形式の場合にエラーを返すことを確認する
  it('名前が大文字で始まる場合にエラーを返す', () => {
    const config: ScaffoldConfig = {
      name: 'Auth',
      tier: 'system',
      apiStyle: 'rest',
      database: 'none',
      description: 'テスト',
      generateClient: false,
    };
    const result = validateConfig(config);
    expect(result.valid).toBe(false);
    expect(result.errors.length).toBeGreaterThan(0);
  });

  // 説明が空の場合にエラーを返すことを確認する
  it('説明が空の場合にエラーを返す', () => {
    const config: ScaffoldConfig = {
      name: 'auth',
      tier: 'system',
      apiStyle: 'rest',
      database: 'none',
      description: '',
      generateClient: false,
    };
    const result = validateConfig(config);
    expect(result.valid).toBe(false);
    expect(result.errors).toContain('description is required');
  });

  // gRPC 使用時に protoPath がない場合にエラーを返すことを確認する
  it('gRPC 使用時に protoPath がない場合にエラーを返す', () => {
    const config: ScaffoldConfig = {
      name: 'auth',
      tier: 'system',
      apiStyle: 'grpc',
      database: 'none',
      description: 'テスト',
      generateClient: false,
    };
    const result = validateConfig(config);
    expect(result.valid).toBe(false);
    expect(result.errors).toContain(
      'protoPath is required when apiStyle includes grpc',
    );
  });

  // both スタイルで protoPath がない場合にエラーを返すことを確認する
  it('both スタイルで protoPath がない場合にエラーを返す', () => {
    const config: ScaffoldConfig = {
      name: 'auth',
      tier: 'system',
      apiStyle: 'both',
      database: 'postgres',
      description: 'テスト',
      generateClient: true,
    };
    const result = validateConfig(config);
    expect(result.valid).toBe(false);
    expect(result.errors).toContain(
      'protoPath is required when apiStyle includes grpc',
    );
  });

  // gRPC 使用時に protoPath があれば成功することを確認する
  it('gRPC 使用時に protoPath があれば valid を返す', () => {
    const config: ScaffoldConfig = {
      name: 'auth',
      tier: 'system',
      apiStyle: 'grpc',
      database: 'none',
      description: 'テスト',
      protoPath: 'proto/auth.proto',
      generateClient: false,
    };
    const result = validateConfig(config);
    expect(result.valid).toBe(true);
  });
});

describe('hasGrpc', () => {
  // grpc と both で true を返すことを確認する
  it('grpc で true を返す', () => {
    expect(hasGrpc('grpc')).toBe(true);
  });

  it('both で true を返す', () => {
    expect(hasGrpc('both')).toBe(true);
  });

  it('rest で false を返す', () => {
    expect(hasGrpc('rest')).toBe(false);
  });
});

describe('hasRest', () => {
  // rest と both で true を返すことを確認する
  it('rest で true を返す', () => {
    expect(hasRest('rest')).toBe(true);
  });

  it('both で true を返す', () => {
    expect(hasRest('both')).toBe(true);
  });

  it('grpc で false を返す', () => {
    expect(hasRest('grpc')).toBe(false);
  });
});

describe('hasDatabase', () => {
  // postgres で true、none で false を返すことを確認する
  it('postgres で true を返す', () => {
    expect(hasDatabase('postgres')).toBe(true);
  });

  it('none で false を返す', () => {
    expect(hasDatabase('none')).toBe(false);
  });
});

describe('toSnakeCase', () => {
  // 各種ケースからスネークケースへの変換を確認する
  it('キャメルケースをスネークケースに変換する', () => {
    expect(toSnakeCase('fooBar')).toBe('foo_bar');
  });

  it('パスカルケースをスネークケースに変換する', () => {
    expect(toSnakeCase('FooBar')).toBe('foo_bar');
  });

  it('ケバブケースをスネークケースに変換する', () => {
    expect(toSnakeCase('foo-bar')).toBe('foo_bar');
  });

  it('既にスネークケースの場合はそのまま返す', () => {
    expect(toSnakeCase('foo_bar')).toBe('foo_bar');
  });
});

describe('toPascalCase', () => {
  // 各種ケースからパスカルケースへの変換を確認する
  it('スネークケースをパスカルケースに変換する', () => {
    expect(toPascalCase('foo_bar')).toBe('FooBar');
  });

  it('ケバブケースをパスカルケースに変換する', () => {
    expect(toPascalCase('foo-bar')).toBe('FooBar');
  });

  it('キャメルケースをパスカルケースに変換する', () => {
    expect(toPascalCase('fooBar')).toBe('FooBar');
  });
});

describe('toKebabCase', () => {
  // 各種ケースからケバブケースへの変換を確認する
  it('キャメルケースをケバブケースに変換する', () => {
    expect(toKebabCase('fooBar')).toBe('foo-bar');
  });

  it('パスカルケースをケバブケースに変換する', () => {
    expect(toKebabCase('FooBar')).toBe('foo-bar');
  });

  it('スネークケースをケバブケースに変換する', () => {
    expect(toKebabCase('foo_bar')).toBe('foo-bar');
  });
});

describe('toCamelCase', () => {
  // 各種ケースからキャメルケースへの変換を確認する
  it('スネークケースをキャメルケースに変換する', () => {
    expect(toCamelCase('foo_bar')).toBe('fooBar');
  });

  it('ケバブケースをキャメルケースに変換する', () => {
    expect(toCamelCase('foo-bar')).toBe('fooBar');
  });

  it('パスカルケースをキャメルケースに変換する', () => {
    expect(toCamelCase('FooBar')).toBe('fooBar');
  });
});

describe('CodegenError', () => {
  // CodegenError のプロパティが正しく設定されることを確認する
  it('code と message が正しく設定される', () => {
    const err = new CodegenError('INVALID_CONFIG', 'config is invalid');
    expect(err.code).toBe('INVALID_CONFIG');
    expect(err.message).toBe('config is invalid');
    expect(err.name).toBe('CodegenError');
  });

  // Error を継承していることを確認する
  it('Error を継承している', () => {
    const err = new CodegenError('TEST', 'test error');
    expect(err).toBeInstanceOf(Error);
    expect(err).toBeInstanceOf(CodegenError);
  });
});

describe('GenerateResult', () => {
  // GenerateResult の型が正しく使えることを確認する
  it('created と skipped を持つオブジェクトを作成できる', () => {
    const result: GenerateResult = {
      created: ['src/main.ts', 'src/config.ts'],
      skipped: ['src/index.ts'],
    };
    expect(result.created).toHaveLength(2);
    expect(result.skipped).toHaveLength(1);
  });
});

describe('ValidationResult', () => {
  // ValidationResult の型が正しく使えることを確認する
  it('valid と errors を持つオブジェクトを作成できる', () => {
    const result: ValidationResult = {
      valid: false,
      errors: ['name is required'],
    };
    expect(result.valid).toBe(false);
    expect(result.errors).toContain('name is required');
  });
});
